File: core/sdk/src/leader_aware.rs
/* Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use iggy_binary_protocol::ClusterClient;
use iggy_common::{
    ClusterMetadata, ClusterNodeRole, ClusterNodeStatus, IggyError, IggyErrorDiscriminants,
};
use std::net::SocketAddr;
use std::str::FromStr;
use tracing::{debug, info, warn};

/// Maximum number of leader redirections to prevent infinite loops
const MAX_LEADER_REDIRECTS: u8 = 3;

/// Check if we need to redirect to leader and return the leader address if redirection is needed
pub async fn check_and_redirect_to_leader<C: ClusterClient>(
    client: &C,
    current_address: &str,
) -> Result<Option<String>, IggyError> {
    debug!("Checking cluster metadata for leader detection");

    match client.get_cluster_metadata().await {
        Ok(metadata) => {
            debug!(
                "Got cluster metadata: {} nodes, cluster: {}",
                metadata.nodes.len(),
                metadata.name
            );
            process_cluster_metadata(&metadata, current_address)
        }
        Err(e) if is_feature_unavailable_error(&e) => {
            debug!("Cluster metadata feature unavailable - server doesn't support clustering");
            Ok(None)
        }
        Err(e) => {
            warn!(
                "Failed to get cluster metadata: {}, connection will continue on server node {}",
                e, current_address
            );
            Ok(None)
        }
    }
}

/// Process cluster metadata and determine if redirection is needed
fn process_cluster_metadata(
    metadata: &ClusterMetadata,
    current_address: &str,
) -> Result<Option<String>, IggyError> {
    let leader = metadata
        .nodes
        .iter()
        .find(|n| n.role == ClusterNodeRole::Leader && n.status == ClusterNodeStatus::Healthy);

    match leader {
        Some(leader_node) => {
            info!(
                "Found leader node: {} at {}",
                leader_node.name, leader_node.address
            );

            if !is_same_address(current_address, &leader_node.address) {
                info!(
                    "Current connection to {} is not the leader, will redirect to {}",
                    current_address, leader_node.address
                );
                Ok(Some(leader_node.address.clone()))
            } else {
                debug!("Already connected to leader at {}", current_address);
                Ok(None)
            }
        }
        None => {
            warn!(
                "No active leader found in cluster metadata, connection will continue on server node {}",
                current_address
            );
            Ok(None)
        }
    }
}

/// Check if two addresses refer to the same endpoint
/// Handles various formats like 127.0.0.1:8090 vs localhost:8090
fn is_same_address(addr1: &str, addr2: &str) -> bool {
    match (parse_address(addr1), parse_address(addr2)) {
        (Some(sock1), Some(sock2)) => sock1.ip() == sock2.ip() && sock1.port() == sock2.port(),
        _ => normalize_address(addr1) == normalize_address(addr2),
    }
}

/// Parse address string to SocketAddr, handling various formats
fn parse_address(addr: &str) -> Option<SocketAddr> {
    if let Ok(socket_addr) = SocketAddr::from_str(addr) {
        return Some(socket_addr);
    }

    let normalized = addr
        .replace("localhost", "127.0.0.1")
        .replace("[::]", "[::1]");

    SocketAddr::from_str(&normalized).ok()
}

/// Normalize address string for comparison
fn normalize_address(addr: &str) -> String {
    addr.to_lowercase()
        .replace("localhost", "127.0.0.1")
        .replace("[::]", "[::1]")
}

/// Check if the error indicates that the feature is unavailable
fn is_feature_unavailable_error(error: &IggyError) -> bool {
    matches!(
        error,
        IggyError::FeatureUnavailable | IggyError::InvalidCommand
    ) || error.as_code() == IggyErrorDiscriminants::FeatureUnavailable as u32
}

/// Struct to track leader redirection state
#[derive(Debug, Clone)]
pub struct LeaderRedirectionState {
    pub redirect_count: u8,
    pub last_leader_address: Option<String>,
}

impl LeaderRedirectionState {
    pub fn new() -> Self {
        Self {
            redirect_count: 0,
            last_leader_address: None,
        }
    }

    pub fn can_redirect(&self) -> bool {
        self.redirect_count < MAX_LEADER_REDIRECTS
    }

    pub fn increment_redirect(&mut self, leader_address: String) {
        self.redirect_count += 1;
        self.last_leader_address = Some(leader_address);
    }

    pub fn reset(&mut self) {
        self.redirect_count = 0;
        self.last_leader_address = None;
    }
}

impl Default for LeaderRedirectionState {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_same_address() {
        assert!(is_same_address("127.0.0.1:8090", "127.0.0.1:8090"));
        assert!(is_same_address("localhost:8090", "127.0.0.1:8090"));
        assert!(!is_same_address("127.0.0.1:8090", "127.0.0.1:8091"));
        assert!(!is_same_address("192.168.1.1:8090", "127.0.0.1:8090"));
    }

    #[test]
    fn test_normalize_address() {
        assert_eq!(normalize_address("localhost:8090"), "127.0.0.1:8090");
        assert_eq!(normalize_address("LOCALHOST:8090"), "127.0.0.1:8090");
        assert_eq!(normalize_address("[::]:8090"), "[::1]:8090");
    }

    #[test]
    fn test_leader_redirection_state() {
        let mut state = LeaderRedirectionState::new();
        assert!(state.can_redirect());
        assert_eq!(state.redirect_count, 0);

        state.increment_redirect("127.0.0.1:8090".to_string());
        assert!(state.can_redirect());
        assert_eq!(state.redirect_count, 1);

        state.increment_redirect("127.0.0.1:8091".to_string());
        state.increment_redirect("127.0.0.1:8092".to_string());
        assert!(!state.can_redirect());
        assert_eq!(state.redirect_count, 3);

        state.reset();
        assert!(state.can_redirect());
        assert_eq!(state.redirect_count, 0);
    }
}

File: foreign/go/internal/util/leader_aware.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package util

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"net"
	"strconv"
	"strings"

	iggcon "github.com/apache/iggy/foreign/go/contracts"
)

// CheckAndRedirectToLeader queries the client for cluster metadata and returns
// an address to redirect to (empty string means no redirection needed).
func CheckAndRedirectToLeader(ctx context.Context, c iggcon.Client, currentAddress string, transport iggcon.Protocol, logger *slog.Logger) (string, error) {
	logger.Debug("Checking cluster metadata for leader detection")

	meta, err := c.GetClusterMetadata(ctx)
	if err != nil {
		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
			return "", err
		}
		logger.Warn(
			"Failed to get cluster metadata, connection will continue on server node",
			"error", err,
			"current_address", currentAddress,
		)
		return "", nil
	}

	logger.Debug(
		"Got cluster metadata",
		"nodes", len(meta.Nodes),
		"cluster", meta.Name,
	)
	return processClusterMetadata(meta, currentAddress, transport, logger)
}

func processClusterMetadata(metadata *iggcon.ClusterMetadata, currentAddress string, transport iggcon.Protocol, logger *slog.Logger) (string, error) {
	if len(metadata.Nodes) == 1 {
		logger.Debug(
			"Single-node cluster detected, no leader redirection needed",
			"node", metadata.Nodes[0].Name,
		)
		return "", nil
	}

	var leader *iggcon.ClusterNode
	for i := range metadata.Nodes {
		node := &metadata.Nodes[i]
		if node.Role == iggcon.RoleLeader && node.Status == iggcon.Healthy {
			leader = node
			break
		}
	}

	if leader == nil {
		logger.Warn(
			"No active leader found in cluster metadata, connection will continue on server node",
			"current_address", currentAddress,
		)
		return "", nil
	}

	var leaderPort uint16
	switch transport {
	case iggcon.Tcp:
		leaderPort = leader.Endpoints.Tcp
	case iggcon.Quic:
		leaderPort = leader.Endpoints.Quic
	case iggcon.Http:
		leaderPort = leader.Endpoints.Http
	case iggcon.WebSocket:
		leaderPort = leader.Endpoints.WebSocket
	default:
		return "", fmt.Errorf("unsupported transport: %v", transport)
	}

	leaderAddress := net.JoinHostPort(leader.IP, strconv.Itoa(int(leaderPort)))
	logger.Debug(
		"Found leader node",
		"leader", leader.Name,
		"leader_address", leaderAddress,
		"transport", transport,
	)

	if !isSameAddress(currentAddress, leaderAddress) {
		logger.Info(
			"Current connection is not the leader, redirecting",
			"current_address", currentAddress,
			"leader_address", leaderAddress,
		)
		return leaderAddress, nil
	}

	logger.Debug("Already connected to leader", "current_address", currentAddress)
	return "", nil
}

// isSameAddress returns true if two addresses refer to the same endpoint.
func isSameAddress(addr1, addr2 string) bool {
	a1 := parseAddress(addr1)
	a2 := parseAddress(addr2)

	if a1 != nil && a2 != nil {
		return a1.IP.Equal(a2.IP) && a1.Port == a2.Port
	}

	return normalizeAddress(addr1) == normalizeAddress(addr2)
}

// parseAddress attempts to parse an address into a *net.TCPAddr.
func parseAddress(addr string) *net.TCPAddr {
	// Try direct parse
	if ta, err := net.ResolveTCPAddr("tcp", addr); err == nil {
		return ta
	}
	// Normalize then try again
	normalized := normalizeAddress(addr)
	if ta, err := net.ResolveTCPAddr("tcp", normalized); err == nil {
		return ta
	}
	return nil
}

// normalizeAddress canonicalizes address strings for fallback comparison.
func normalizeAddress(addr string) string {
	out := strings.ToLower(addr)
	out = strings.ReplaceAll(out, "localhost", "127.0.0.1")
	out = strings.ReplaceAll(out, "[::]", "[::1]")
	return out
}
